### DISCUSSION TITLE
How I solved the problem of `cfg/cfg_attr` expansion in proc macro attributes

### DISCUSSION THREAD

**Discussion by Veetaha** [elastio/bon#124] [Show and tell]
This is in response to the [reddit comment](https://www.reddit.com/r/rust/comments/1fc0ai7/comment/lmg3vlq/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) by @jplatte.

When developing a hack for `cfg/cfg_attr` handing in proc macro attributes, I tried almost everything possible. I even tried generating a dummy struct with a `derive()` (as suggested in @jplatte's Reddit comment). It looked smth like this:

```rust
#[derive(bon::__ExpandCfgs)]
#[expand_cfgs(
    // Outputs a call to the proc macro attribute specified in this parameter
    callback = bon::builder,

    // Passes through these params at the end of `callback(...)`. Here
    // should be the parameters passed to the `#[builder(...)]` by the user
    // that are required for its business logic
    params(/* */),

    // The item the proc-macro-attribute was placed on top of
    item(fn greet(/* */) {/* */} )
)]
// Add a `cfg_attr` for every cfg expression. They will be evaluated
// by the compiler automatically before the derive(bon::__ExpandCfgs) runs.
// This way __ExpandCfgs can see which cfgs evaluated to `true`, and consider
// all other cfgs to be `false`.
#[cfg_attr(feature = "foo", expand_cfgs(cfg(feature = "foo")))]
#[cfg_attr(feature = "bar", expand_cfgs(cfg(feature = "bar")))]
struct DummyStruct;
```

Now let's suppose that only `feature = "bar"` cfg evaluates to `true`.

In this case, the logic of `derive(bon::__ExpandCfgs)` would expand to the following code:

```rust
                   // v- here goes the comma-separated list of all cfg predicates that evaluate to `true`
#[bon::builder(__cfgs(feature = "foo") /* params */)]
/* item */
```

Then `bon::builder` would parse the special `__cfgs(...)` syntax in front of its parameters to collect the cfgs expressions that evaluate to `true`. Then there is a bit of code that traverses the function or impl block that `bon::builder` was placed on and evaluates the `cfg/cfg_attr` manually (it basically reimplements these attrs inside of `bon::builder`). Yeah... that's unfortunate:
https://github.com/elastio/bon/blob/3217b4b0349f03f0b2a5853310f420c5b8b005a7/bon-macros/src/normalization/cfg/mod.rs#L191-L243

I implemented this logic, tested it and it indeed worked! Yes, but... it didn't work well with Rust Analyzer. The entire function body with the `#[bon::builder]` annotation was completely excluded from RA's analysis. No code completions and on-hover hints worked in it. Why? I'm not entirely sure, but I think what causes this is that function item is passed as an argument to the derive macro like `#[expand_cfgs(item(fn greet(/**/) {/* */}))]`. This somehow makes RA lose the span information, I guess.

I even tried a crazy hack of passing the function item in a nested expression inside of the enum's discriminant value expression. However, I stumbled with a bug in rustc, in that it invalidly expands such code (see https://github.com/rust-lang/rust/issues/130004). I suppose no one ever tried doing such a crazy hack so no one ever caught such a bug.

So, in the end, I didn't use the `derive` syntax approach to evaluate the cfgs, instead I used the `macro_rules!` that can be observed here:
https://github.com/elastio/bon/blob/3217b4b0349f03f0b2a5853310f420c5b8b005a7/bon/src/private/mod.rs#L85-L188)

So the `#[bon::builder]` macro generates a call to this `macro_rules` which basically does the similar job of expanding the cfgs, collecting the results and passing them as `__cfg(...)` in front of the callback proc macro attribute parameters. This, surprisingly works well with RA. RA can even recursively expand the macro with its "Expand macro recursively" feature.

Note that both of these solutions suffer from a very silly problem that you need to generate unique names for the struct under the derive (if you use the derive approach) or the symbols created by `use` statements in the `macro_rules!` approach current used in `bon`. You probably don't want macros to generate non-deterministic random identifiers on each rerun to avoid any unknown headaches that this may bring, so there is a bit of heuristic solution to this problem implemented here:

https://github.com/elastio/bon/blob/3217b4b0349f03f0b2a5853310f420c5b8b005a7/bon-macros/src/normalization/cfg/mod.rs#L129-L189

**Comment by jplatte** [comment]
I think the `__ExpandCfgs` approach is very different from what I suggested on reddit, I don't fully understand it but it seems to be trying to get the info of which cfgs evaluate to true / false into the attribute parameters, such that the attribute can then evaluate cfgs it sees.

That is not what I was talking about: Rather, the idea would be that `#[bon::builder]` could delegate most of its work (of generating impl blocks) to a derive macro¹, which would make rustc remove all `cfg(<evals to false>)` arguments for you. Assuming `bon` were to generate untyped builders with everything just being `Option`... I know it's not nearly this simple, and maybe that actually means this idea doesn't work, but maybe it does:

```rust
// user written code:
#[bon::builder]
fn greet(
    #[cfg(all())] name: &str,
    #[cfg(any())] level: Option<u32>,
) -> String {
    format!("Hello {name}!")
}

// expands to
fn greet() -> GreetBuilder {
    GreetBuilder {
        #[cfg(all())]
        name: None,
        #[cfg(any())]
        level: None,
    }
}

#[derive(bon::Builder)]
#[bon(__private_fn_builder)]
struct GreetBuilder<'a> {
    #[cfg(all())]
    name: Option<&'a str>,
    #[cfg(any(())]
    level: Option<Option<u32>>,
}

// what the Builder derive sees
#[bon(__private_fn_builder)]
struct GreetBuilder<'a> {
    name: Option<&'a str>,
}

// so it can generate an `impl<'a> GreetBuilder<'a>`
// knowing that there is (only) a name field
```

Am I talking nonsense? I applied this pattern successfully before so I know the general idea works, but maybe it's not applicable here.

¹ either always, or only if there are any `cfg`s / `cfg_attr`s in the input as an optimization

**Reply by Veetaha** [reply]
Aha, I see what's your suggestion now. It makes sense indeed, although it would be rather complicated to apply to `bon` because the current generated builder looks like this:

```rust
type GreetBuilderInitialState = (
    Unset<Required>,
    Unset<Optional>,
);

struct GreetBuilder<State = GreetBuilderInitialState> {
    __private_members: State,
    /* other unimportant stuff */
}
```

Where the `State` is a type state expressed as a tuple with one element per each function argument (and attributes aren't supported inside of tuple types). It's theoretically possible to update the codegen to have a field per member in the builder, but that leads to other complications (and `bon` already was there).

However, the idea looks nice, I'll keep it in mind! If I find other limitations with the current approach I'll consider it, but yeah, it's not easily applicable in the current design of the builder struct. Anyway thank you, for the suggestion🐱

**Reply by jplatte** [reply]
Huh, I didn't know that attributes are forbidden on tuple fields!
I wonder why that is, considering that [it works with tuple structs](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8cdd80798432275dd35d494f943c0f65).

**Reply by Veetaha** [reply]
Yeah, there are a lot of places in Rust where attributes aren't supported (yet?). It's really easy to check which context supports attributes by just reviewing the `syn`'s AST models. For example the [`TypeTuple`](https://docs.rs/syn/latest/syn/struct.TypeTuple.html) in question here doesn't have them, and I think attributes are not allowed inside of basically any type.

The [`Field`](https://docs.rs/syn/latest/syn/struct.Field.html) of the tuple struct does have the attributes.
